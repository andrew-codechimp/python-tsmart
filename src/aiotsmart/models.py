"""Models for TSmart."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum


class Mode(IntEnum):
    """TSmart Modes."""

    MANUAL = 0x00
    ECO = 0x01
    SMART = 0x02
    TIMER = 0x03
    TRAVEL = 0x04
    BOOST = 0x05
    LIMITED = 0x21
    CRITICAL = 0x22


@dataclass
class DiscoveredDevice:
    """Discovery model."""

    ip_address: str
    device_id: str
    device_name: str


@dataclass
class Configuration:
    """Configuration model."""

    device_id: str
    device_name: str
    firmware_version: str
    firmware_name: str
    raw_response: bytes


@dataclass
class Status:
    """Status model."""

    power: bool
    setpoint: str
    mode: Mode
    temperature_high: str
    temperature_low: str
    temperature_average: str
    relay: bool
    error_e01: bool
    error_e02: bool
    error_e03: bool
    error_e04: bool
    error_e05: bool
    error_w01: bool
    error_w02: bool
    error_w03: bool
    raw_response: bytes

    @property
    def has_error(self) -> bool:
        """Is there any error."""
        return (
            self.error_e01
            or self.error_e02
            or self.error_e03
            or self.error_e04
            or self.error_e05
            or self.error_w01
            or self.error_w02
            or self.error_w03
        )
