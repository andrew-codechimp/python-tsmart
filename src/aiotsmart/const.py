"""Constants for tsmart."""

MESSAGE_HEADER = "=BBBB"
UDP_PORT = 1337
