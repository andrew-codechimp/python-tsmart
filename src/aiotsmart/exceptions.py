"""Asynchronous Python client for TSmart."""


class TSmartError(Exception):
    """Generic exception."""


class TSmartCancelledError(TSmartError):
    """TSmart cancelled exception."""


class TSmartTimeoutError(TSmartError):
    """TSmart timeout exception."""


class TSmartBadResponseError(TSmartError):
    """TSmart bad response exception."""
