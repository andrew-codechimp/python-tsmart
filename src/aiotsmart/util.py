"""Utility methods."""

from __future__ import annotations


def validate_checksum(data: bytes) -> bool:
    """Validate the checksum."""

    t = 0
    for b in data[:-1]:
        t = t ^ b
    return t ^ 0x55 == data[-1]
