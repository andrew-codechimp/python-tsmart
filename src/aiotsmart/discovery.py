"""TSmart Discovery."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
import logging
import socket
import struct
from typing import Any, Callable, Self

from aiotsmart.models import DiscoveredDevice
from aiotsmart.util import validate_checksum

from .const import MESSAGE_HEADER, UDP_PORT

DISCOVERY_INTERVAL = 2  # seconds
DISCOVERY_MESSAGE = struct.pack(MESSAGE_HEADER, 0x01, 0, 0, 0x01 ^ 0x55)
BROADCAST_ADDR = ("255.255.255.255", UDP_PORT)

_LOGGER = logging.getLogger(__name__)

SHARED_LIST: list[DiscoveredDevice] = []


def _unpack_discovery_response(
    data: bytes, addr: tuple[str, int]
) -> dict[str, str] | None:
    """Return dict of unpacked responses from TSmart Immersion Heater."""
    response_struct = struct.Struct("=BBBHL32sBB")

    remote_addr_ip_address = addr[0]

    result = {"ip_address": addr[0]}

    if len(data) == len(DISCOVERY_MESSAGE):
        # Got our own broadcast
        return None

    if len(data) != response_struct.size:
        _LOGGER.debug(
            "Unexpected packet length (got: %d, expected: %d)"
            % (len(data), response_struct.size)
        )
        return None

    if data[0] == 0:
        _LOGGER.debug("Got error response (code %d)" % (data[0]))
        return None

    if data[0] != DISCOVERY_MESSAGE[0]:
        _LOGGER.debug(
            "Unexpected response type (%02X %02X %02X)" % (data[0], data[1], data[2])
        )
        return None

    if not validate_checksum(data):
        _LOGGER.debug("Received packet checksum failed")
        return None

    _LOGGER.debug("Got response from %s", remote_addr_ip_address)

    # pylint:disable=unused-variable
    (
        cmd,
        sub,
        sub2,
        device_type,
        device_id,
        name,
        tz,
        checksum,
    ) = response_struct.unpack(data)

    result["device_name"] = name.decode("utf-8").split("\x00")[0]
    result["device_id"] = f"{device_id:04X}"
    _LOGGER.info("Discovered %s %s" % (result["device_id"], result["device_name"]))

    return result


class DiscoveryProtocol(asyncio.DatagramProtocol):
    """Protocol to send discovery request and receive responses."""

    def __init__(self, callback: Callable[[DiscoveredDevice], None]) -> None:
        """Initialize with callback function."""
        self.transport = None
        self.callback = callback

    def connection_made(self, transport: Any) -> None:
        """Connect to transport."""
        self.transport = transport

    def datagram_received(self, data: bytes, addr: tuple[str | Any, int]) -> None:
        """Test if responder is a TSmart Immersion Heater."""
        _LOGGER.debug("Received discovery response from %s", addr)
        response = _unpack_discovery_response(data, addr)
        if response:
            if (
                "ip_address" not in response
                or "device_name" not in response
                or "device_id" not in response
            ):
                _LOGGER.info(
                    "TSmart discovery response %s does not contain enough information to connect",
                    response,
                )
            if callable(self.callback):
                result = self.callback(
                    DiscoveredDevice(
                        response["ip_address"],
                        response["device_id"],
                        response["device_name"],
                    )
                )
                if asyncio.iscoroutine(result):
                    asyncio.create_task(result)


@dataclass
class TSmartDiscovery:
    """TSmart Discovery."""

    _discovered_devices: list[DiscoveredDevice] = field(
        default_factory=lambda: SHARED_LIST
    )

    def _device_discovered(self, device: DiscoveredDevice) -> None:
        """Add device to discover list if new."""

        matched_device: DiscoveredDevice | None = next(
            (x for x in self._discovered_devices if x.ip_address == device.ip_address),
            None,
        )

        if not matched_device:
            self._discovered_devices.append(device)

    async def discover(self) -> list[DiscoveredDevice]:
        """Broadcast discovery packet and return a list of discovered devices."""
        loop = asyncio.get_running_loop()

        sock = socket.socket(
            socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP
        )  # Internet, UDP

        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)

        sock.bind(("", UDP_PORT))

        transport, _ = await loop.create_datagram_endpoint(
            lambda: DiscoveryProtocol(self._device_discovered),
            sock=sock,
        )

        try:
            for _ in range(2):
                _LOGGER.debug("Sending discovery message.")
                transport.sendto(DISCOVERY_MESSAGE, BROADCAST_ADDR)
                await asyncio.sleep(DISCOVERY_INTERVAL)

        except asyncio.CancelledError:
            _LOGGER.debug("Cancelling TSmart discovery task")
            transport.close()

        finally:
            transport.close()

        return self._discovered_devices

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The TSmartClient object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
