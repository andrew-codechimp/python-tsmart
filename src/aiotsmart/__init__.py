"""Asynchronous Python client for TSmart."""

from aiotsmart.exceptions import (
    TSmartConnectionError,
    TSmartError,
    TSmartNoResponseError,
    TSmartNotFoundError,
    TSmartTimeoutError,
)
from aiotsmart.discovery import TSmartDiscovery
from aiotsmart.models import Configuration, DiscoveredDevice, Mode, Status
from aiotsmart.tsmart import TSmartClient

__all__ = [
    "TSmartDiscovery",
    "TSmartNoResponseError",
    "TSmartConnectionError",
    "TSmartError",
    "TSmartNotFoundError",
    "TSmartTimeoutError",
    "TSmartClient",
    "Configuration",
    "DiscoveredDevice",
    "Status",
    "Mode",
]
