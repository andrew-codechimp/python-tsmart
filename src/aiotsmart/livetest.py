"""Tests."""

import asyncio

from aiotsmart import TSmartClient, Mode


client = TSmartClient("192.168.1.35")


async def do_stuff():
    """Test."""

    # discovery = TSmartDiscovery()
    # devices = await discovery.discover()
    # print(devices)

    # configuration = await client.configuration_read()
    # print(configuration)

    status = await client.control_read()
    print(status)

    print(status.has_error)

    # try:
    #     status = await client.control_read()
    #     pprint(vars(status))
    # except TSmartError as exception:
    #     print(repr(exception))

    # await client.control_write(power=True, mode=Mode.MANUAL, setpoint=15)

    # pprint(vars(client))


asyncio.run(do_stuff())
