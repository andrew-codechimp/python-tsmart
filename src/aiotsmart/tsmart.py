"""Homeassistant Client."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
import logging
import socket
import struct
from typing import Any, Self, Callable

from aiotsmart.exceptions import (
    TSmartBadResponseError,
    TSmartCancelledError,
    TSmartTimeoutError,
)
from aiotsmart.models import Configuration, Mode, Status
from aiotsmart.util import validate_checksum, add_checksum

from .const import MESSAGE_HEADER, UDP_PORT

_LOGGER = logging.getLogger(__name__)
TIMEOUT = 5  # seconds


# pylint:disable=too-many-locals
def _unpack_configuration_response(
    request: bytearray, data: bytes
) -> Configuration | None:
    """Return unpacked configuration response from TSmart Immersion Heater."""
    response_struct = struct.Struct("=BBBHL32sBBBBB32s28s32s64s124s")

    if len(data) != response_struct.size:
        raise TSmartBadResponseError(
            "Unexpected packet length (got: %d, expected: %d)"
            % (len(data), response_struct.size)
        )

    if data[0] == 0:
        raise TSmartBadResponseError("Got error response (code %d)" % (data[0]))

    if data[0] != request[0]:
        raise TSmartBadResponseError(
            "Unexpected response type (%02X %02X %02X)" % (data[0], data[1], data[2])
        )

    if not validate_checksum(data):
        raise TSmartBadResponseError("Received packet checksum failed")

    # pylint:disable=unused-variable
    (
        cmd,
        sub,
        sub2,
        device_type,
        device_id,
        device_name,
        tz,
        userbin,
        firmware_version_major,
        firmware_version_minor,
        firmware_version_deployment,
        firmware_name,
        legacy,
        wifi_ssid,
        wifi_password,
        unused,
    ) = response_struct.unpack(data)

    configuration = Configuration(
        device_id=f"{device_id:04X}",
        device_name=device_name.decode("utf-8").split("\x00")[0],
        firmware_version=f"{firmware_version_major}.{firmware_version_minor}.{firmware_version_deployment}",
        firmware_name=firmware_name.decode("utf-8").split("\x00")[0],
        raw_response=data,
    )
    _LOGGER.info(
        "Configuration received %s %s"
        % (configuration.device_id, configuration.device_name)
    )

    return configuration


# pylint:disable=too-many-locals
def _unpack_control_read_response(request: bytearray, data: bytes) -> Status | None:
    """Return unpacked control read response from TSmart Immersion Heater."""
    response_struct = struct.Struct("=BBBBHBHBBH16sB")

    if len(data) != response_struct.size:
        raise TSmartBadResponseError(
            "Unexpected packet length (got: %d, expected: %d)"
            % (len(data), response_struct.size)
        )

    if data[0] == 0:
        raise TSmartBadResponseError(("Got error response (code %d)" % (data[0])))

    if data[0] != request[0]:
        raise TSmartBadResponseError(
            "Unexpected response type (%02X %02X %02X)" % (data[0], data[1], data[2])
        )

    if not validate_checksum(data):
        raise TSmartBadResponseError("Received packet checksum failed")

    # pylint:disable=unused-variable
    (
        cmd,
        sub,
        sub2,
        power,
        setpoint,
        mode,
        t_high,
        relay,
        smart_state,
        t_low,
        error_buffer,
        checksum,
    ) = response_struct.unpack(data)

    status = Status(
        power=bool(power),
        setpoint=setpoint / 10,
        mode=Mode(mode),
        temperature_high=t_high / 10,
        temperature_low=t_low / 10,
        temperature_average=(t_high + t_low) / 20,
        relay=bool(relay),
        error_e01=(error_buffer[0] >> 7) & 1 == 1,
        error_e02=(error_buffer[2] >> 7) & 1 == 1,
        error_e03=(error_buffer[4] >> 7) & 1 == 1,
        error_e04=(error_buffer[6] >> 7) & 1 == 1,
        error_w01=(error_buffer[8] >> 7) & 1 == 1,
        error_w02=(error_buffer[10] >> 7) & 1 == 1,
        error_w03=(error_buffer[12] >> 7) & 1 == 1,
        error_e05=(error_buffer[14] >> 7) & 1 == 1,
        raw_response=data,
    )
    return status


# pylint:disable=too-many-locals
def _unpack_control_write_response(_: bytearray, data: bytes) -> None:
    """Return unpacked control write response from TSmart Immersion Heater."""

    if data != b"\xf2\x00\x00\xa7":
        raise TSmartBadResponseError


class TsmartProtocol(asyncio.DatagramProtocol):
    """Protocol to send request and receive responses."""

    def __init__(
        self, request: bytearray, unpack_function: Callable[[bytearray, bytes], Any]
    ) -> None:
        """Initialize with callback function."""
        self.request = request
        self.unpack_function = unpack_function
        self.done = asyncio.get_running_loop().create_future()

    def datagram_received(self, data: bytes, addr: tuple[str | Any, int]) -> None:
        """Test if responder is a TSmart Immersion Heater."""
        _LOGGER.debug("Received configuration response from %s", addr)
        response = self.unpack_function(self.request, data)

        self.done.set_result(response)


@dataclass
class TSmartClient:
    """TSmart Client."""

    ip_address: str

    def create_socket(self) -> socket.socket:
        """Create a UDP socket."""
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # Internet, UDP

        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.bind(("", 1337))
        sock.connect((self.ip_address, UDP_PORT))
        return sock

    async def configuration_read(self) -> Configuration:
        """Get configuration from immersion heater."""

        loop = asyncio.get_running_loop()

        sock = self.create_socket()

        request = struct.pack(MESSAGE_HEADER, 0x21, 0, 0, 0)
        request_checksum = add_checksum(request)

        transport, protocol = await loop.create_datagram_endpoint(
            lambda: TsmartProtocol(request_checksum, _unpack_configuration_response),
            sock=sock,
        )

        try:
            _LOGGER.debug("Sending configuration message.")
            async with asyncio.timeout(TIMEOUT):
                transport.sendto(request_checksum, (self.ip_address, UDP_PORT))
                configuration: Configuration = await protocol.done
        except asyncio.TimeoutError as ex:
            raise TSmartTimeoutError() from ex
        except asyncio.CancelledError as ex:
            raise TSmartCancelledError() from ex

        finally:
            transport.close()
            sock.close()

        _LOGGER.info("Received configuration from %s" % self.ip_address)

        return configuration

    # pylint:disable=too-many-locals
    async def control_read(self) -> Status:
        """Get status from the immersion heater."""

        loop = asyncio.get_running_loop()

        sock = self.create_socket()

        request = struct.pack(MESSAGE_HEADER, 0xF1, 0, 0, 0)
        request_checksum = add_checksum(request)

        transport, protocol = await loop.create_datagram_endpoint(
            lambda: TsmartProtocol(request_checksum, _unpack_control_read_response),
            sock=sock,
        )

        try:
            _LOGGER.debug("Sending control message.")
            async with asyncio.timeout(TIMEOUT):
                transport.sendto(request_checksum, (self.ip_address, UDP_PORT))
                status: Status = await protocol.done
        except asyncio.TimeoutError as ex:
            raise TSmartTimeoutError() from ex

        except asyncio.CancelledError as ex:
            raise TSmartCancelledError() from ex

        finally:
            transport.close()
            sock.close()

        _LOGGER.info("Received control from %s" % self.ip_address)

        return status

    async def control_write(self, power: bool, mode: Mode, setpoint: int) -> None:
        """Set the immersion heater."""

        _LOGGER.info("Control set %d %d %0.2f" % (power, mode, setpoint))

        loop = asyncio.get_running_loop()

        sock = self.create_socket()

        request = struct.pack(
            "=BBBBHBB", 0xF2, 0, 0, 1 if power else 0, setpoint * 10, mode, 0
        )
        request_checksum = add_checksum(request)

        transport, protocol = await loop.create_datagram_endpoint(
            lambda: TsmartProtocol(request_checksum, _unpack_control_write_response),
            sock=sock,
        )

        try:
            _LOGGER.debug("Sending control message.")
            async with asyncio.timeout(TIMEOUT):
                transport.sendto(request_checksum, (self.ip_address, UDP_PORT))
                await protocol.done
        except asyncio.TimeoutError as ex:
            raise TSmartTimeoutError() from ex

        except asyncio.CancelledError as ex:
            raise TSmartCancelledError() from ex

        finally:
            transport.close()
            sock.close()

        _LOGGER.info("Received control from %s" % self.ip_address)

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The TSmartClient object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
