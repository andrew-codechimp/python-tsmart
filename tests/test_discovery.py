"""Asynchronous Python client for TSmart."""

from __future__ import annotations

from typing import TYPE_CHECKING
from unittest.mock import patch

import pytest

import aiotsmart
import aiotsmart.discovery

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion

ADDR = ("192.168.1.1", 1337)
DATA = b"\x01\x00\x00 \x00\r*\x9b\x00TESLA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00d\xe3"
BAD_DATA = b"\x05\x00\x09 \x09\r*\x9b\x00TESLA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00d\xe3"
RESPONSE = {"ip_address": "192.168.1.35", "device_id": "9B2A0D", "device_name": "TESLA"}

# pylint: disable=C0103
# All test coroutines will be treated as marked.
pytestmark = pytest.mark.asyncio


async def test_discovery_unpack() -> None:
    """Test device discovery unpack."""

    # pylint:disable=protected-access
    assert aiotsmart.discovery._unpack_discovery_response(DATA, ADDR)
    assert not aiotsmart.discovery._unpack_discovery_response(BAD_DATA, ADDR)


async def test_discovery(
    snapshot: SnapshotAssertion,
) -> None:
    """Test discovery."""
    with patch("aiotsmart.discovery._unpack_discovery_response", return_value=RESPONSE):
        assert aiotsmart.TSmartDiscovery.discover == snapshot
    #     protocol = pysqueezebox.discovery.ServerDiscoveryProtocol(callback)
    #     async_protocol = pysqueezebox.discovery.ServerDiscoveryProtocol(async_callback)
    #     protocol.datagram_received(ADDR, DATA)
    #     async_protocol.datagram_received(ADDR, DATA)

    # callback.assert_called_once()
    # async_callback.assert_called_once()

    # with patch(
    #     "aiotsmart.TSmartDiscovery._request",
    #     return_value=b"!\x00\x00 \x00\r*\x9b\x00TESLA\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00d\x00\x01\t`Boiler\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\xff\xff\x01\x01\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x00\x00\xff\xff\xff\xff\x00\x00wifissid\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00wifipassword\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xd0!\xf9\xb1\xd6QTESLA_74BA1D\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0c\x01\xa7\x1e\x00\x00\x00\x00\x00\x04d\x00\t\x00\x00\x00\x01\x00\x00\x00n",
    # ):
    #     assert await tsmart_client.configuration_read() == snapshot
