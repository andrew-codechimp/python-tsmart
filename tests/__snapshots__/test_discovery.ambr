# serializer version: 1
# name: test_discovery
  "TSmartDiscovery.discover(self) -> 'list[DiscoveredDevice]'"
# ---
